File: RetirementSavingsCalculator/src/Strategies.Core/Models/WithdrawalStrategy.cs
namespace Strategies.Core.Models;

public class WithdrawalStrategy
{
    public Guid WithdrawalStrategyId { get; set; }
    public Guid TenantId { get; set; }
    public Guid RetirementScenarioId { get; set; }
    public string Name { get; set; } = string.Empty;
    public decimal WithdrawalRate { get; set; }
    public decimal AnnualWithdrawalAmount { get; set; }
    public bool AdjustForInflation { get; set; }
    public decimal? MinimumBalance { get; set; }
    public WithdrawalStrategyType StrategyType { get; set; }
    public string? Notes { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
}

public enum WithdrawalStrategyType
{
    FixedAmount = 0,
    PercentageBased = 1,
    Dynamic = 2,
    RequiredMinimumDistribution = 3,
}

File: RetirementSavingsCalculator/src/RetirementSavingsCalculator.WebApp/projects/retirement-savings-calculator-shared/src/lib/models/withdrawal-strategy.ts
export enum WithdrawalStrategyType {
  FixedAmount = 0,
  PercentageBased = 1,
  Dynamic = 2,
  RequiredMinimumDistribution = 3
}

export const WithdrawalStrategyTypeLabels: Record<WithdrawalStrategyType, string> = {
  [WithdrawalStrategyType.FixedAmount]: 'Fixed Amount',
  [WithdrawalStrategyType.PercentageBased]: 'Percentage Based',
  [WithdrawalStrategyType.Dynamic]: 'Dynamic',
  [WithdrawalStrategyType.RequiredMinimumDistribution]: 'Required Minimum Distribution'
};

export interface WithdrawalStrategy {
  withdrawalStrategyId: string;
  retirementScenarioId: string;
  name: string;
  withdrawalRate: number;
  annualWithdrawalAmount: number;
  adjustForInflation: boolean;
  minimumBalance?: number;
  strategyType: WithdrawalStrategyType;
  notes?: string;
}

export interface CreateWithdrawalStrategy {
  retirementScenarioId: string;
  name: string;
  withdrawalRate: number;
  annualWithdrawalAmount: number;
  adjustForInflation: boolean;
  minimumBalance?: number;
  strategyType: WithdrawalStrategyType;
  notes?: string;
}

export interface UpdateWithdrawalStrategy {
  withdrawalStrategyId: string;
  name: string;
  withdrawalRate: number;
  annualWithdrawalAmount: number;
  adjustForInflation: boolean;
  minimumBalance?: number;
  strategyType: WithdrawalStrategyType;
  notes?: string;
}
